Shipping Labels: Fix issue purchasing labels when navigating from a purchased label

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: WOOMOB-347
Note: Please review #15528 before this one.

<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Currently, there are a few issues regarding origin and destination addresses when multiple shipments exist:
- When navigating to the shipping label creation form from a purchased label on the order details screen, loading shipping rates would fail with an error message indicating a missing destination address. ~~The error is due to both origin and destination addresses are being passed as publishers from a `@Published` property, which doesn't emit the current value upon subscription.~~ The error is due to origin address not being loaded when opening a purchased label.
- When updating the origin and destination addresses of an unfulfilled shipment, the addresses of purchased labels are updated accordingly.

The solution:
- ~~Switch `selectedOriginAddress` and `destinationAddress` in `WooShippingCreateLabelsViewModel` to `CurrentValueSubject` to enable emitting current value upon subscription.~~
- Enable loading origin addresses when there are unfulfilled shipments.
- Updated `originAddressLines` and `destinationAddressLines` to ensure that purchased labels always display correct addresses.
- Refactoring: updated the type of `originAddress` in `WooShippingShipmentDetailsViewModel` to simplify the logic.


## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Log in to a test store with Woo Shipping plugin set up.
2. Navigate to the Orders tab and select a completed order with several physical products.
3. Select Create shipping labels > Split shipments > Move items to different shipments.
4. Proceed to purchase a label for a shipment.
5. Navigate back to the order details screen after purchase completes.
6. Select View purchased shipping label for the newly purchased label.
7. On the shipping label form, switch to the tab of an unfulfilled shipment.
8. Select a package and confirm that shipping rates can be loaded successfully.
9. Tap on Shipment details sheet of the unfulfilled shipment and update either the origin or destination addresses or both.
10. Switch to the tab of the fulfilled shipment > Tap Shipment details and confirm that the origin and destination addresses are still correct.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
- Manually tested and confirmed the above cases on simulator iPhone 16 Pro iOS 18.2.
- Added unit tests to verify `originAddressLines` and `destinationAddressLines`.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/f9fbf784-bccb-4aee-bb45-77610ac5dd0d



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

itsmeichigo committed:
Load origin address list when there are unfulfilled shipments

itsmeichigo committed:
Display correct origin and destination addresses for purchased labels

itsmeichigo committed:
Use CurrentValueSubject for selectedOriginAddress and destinationAddress

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Fix unit test build failures

itsmeichigo committed:
Add unit tests for originAddressLines and destinationAddressLines

RafaelKayumov commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShippingCreateLabelsViewModel.swift:
```diff
@@ -75,30 +75,38 @@ final class WooShippingCreateLabelsViewModel: ObservableObject {
     /// Address to ship from (store address).
-    @Published private var selectedOriginAddress: WooShippingOriginAddress?
+    private let selectedOriginAddress = CurrentValueSubject<WooShippingOriginAddress?, Never>(nil)
 
     /// Address to ship to (customer address),
-    @Published private var destinationAddress: WooShippingAddress?
+    private let destinationAddress = CurrentValueSubject<WooShippingAddress?, Never>(nil)
```
Could you please give more details about benefits of `CurrentValueSubject` against `@Published` in terms of emitting current value upon subscription? I thought they behaved the same way.

itsmeichigo replied:
I'd say `@Published` is similar to `CurrentValueSubject` in the sense that you can get the current state when you access it directly through the wrapped property. But it's similar to `PassthroughSubject` when used as a publisher - it will only emit new values, not the value set before the subscription.

I cannot find a clear documentation about this, but you can test this code in a playground:

```Swift
import UIKit
import Combine

final class PublisherTest {
    @Published var publishedNumber: Int?
    let currentValueSubjectNumber = CurrentValueSubject<Int?, Never>(nil)

    init(number: Int?) {
        self.publishedNumber = publishedNumber
        currentValueSubjectNumber.send(number)
    }
}

final class ObserverTest {
    let test = PublisherTest(number: 1)
    @Published var publishedNumberObserver: Int?
    @Published var currentValueSubjectObserver: Int?

    func performObservation() {
        test.$publishedNumber
            .assign(to: &$publishedNumberObserver)

        test.currentValueSubjectNumber
            .assign(to: &$currentValueSubjectObserver)
    }
}

let observerTest = ObserverTest()
observerTest.performObservation()
observerTest.publishedNumberObserver
observerTest.currentValueSubjectObserver
```

The result should show that `observerTest.publishedNumberObserver` is nil and `observerTest.currentValueSubjectObserver` is 1.

RafaelKayumov replied:
Thx for the example

RafaelKayumov reviewed (commented):
Tested with listed steps on iPhone 16 (18.2) Simulator - works as described. Shipping rates do appear. Changing addresses on unfulfilled shipment doesn't affect the one with shipping label purchased.

Not sure if this is a known issue - I noticed a weird behaviour where it was possible to move items from unfulfilled shipments to the fulfilled one with a label purchased. Following navigations recorded on the video lead to corruption of destination addresses.

https://github.com/user-attachments/assets/64078094-e88b-400c-bebb-160fd11f959b

itsmeichigo committed:
Revert changes to @Published properties for simplicity

itsmeichigo commented:
Thanks @RafaelKayumov for the review!

> Not sure if this is a known issue - I noticed a weird behaviour where it was possible to move items from unfulfilled shipments to the fulfilled one with a label purchased. Following navigations recorded on the video lead to corruption of destination addresses.

Yes, this has been raised in #15483 and is handled in #15543.

Thanks to your question regarding the use of `CurrentValueSubject`, I noticed that the change is not necessary anymore since I enabled [loading origin address](https://github.com/woocommerce/woocommerce-ios/pull/15537/commits/7296edc59fb7c24591c158eb2e9518471b9e6ad1). I reverted the changes in 29bc3741d792cc7b1054e836a6a68a5d3a799d0f and updated the PR description accordingly.

This PR is ready for another look.
